File: bin/website-epub.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '.common/script_base'
require_relative '.common/services/browser_service'
require_relative '.common/services/page_fetcher'
require_relative '.common/services/epub_generator'
require_relative '.common/concerns/article_detector'

# Website EPUB Creator - Scrape website and extract all URLs
class WebsiteEpub < ScriptBase
  def script_emoji
    '🌐'
  end

  def script_title
    'Website EPUB Creator'
  end

  def script_description
    'Scrapes a website using browser automation and extracts all HTTP/HTTPS URLs'
  end

  def script_arguments
    '<url>'
  end

  def add_custom_options(opts)
    opts.on('--max-clicks COUNT', Integer, 'Maximum read more clicks (default: 5)') do |count|
      @options[:max_clicks] = count
    end

    opts.on('--[no-]headless', 'Run browser in headless mode (default: true)') do |headless|
      @options[:headless] = headless
    end

    opts.on('--timeout SECONDS', Integer, 'Browser timeout in seconds (default: 30)') do |timeout|
      @options[:timeout] = timeout
    end

    opts.on('--wait SECONDS', Integer, 'Wait time for page load (default: 3)') do |wait|
      @options[:wait] = wait
    end

    opts.on('--[no-]javascript', 'Use JavaScript for article fetching (default: false)') do |js|
      @options[:use_javascript] = js
    end

    opts.on('--[no-]cache', 'Enable page caching (default: true)') do |cache|
      @options[:cache_enabled] = cache
    end

    opts.on('--articles-only', 'Only process URLs that are detected as articles') do
      @options[:articles_only] = true
    end

    opts.on('--[no-]create-epub', 'Create EPUB file from detected articles (default: false)') do |create|
      @options[:create_epub] = create
    end

    opts.on('--epub-title TITLE', 'Custom title for the EPUB') do |title|
      @options[:epub_title] = title
    end

    opts.on('--epub-author AUTHOR', 'Custom author for the EPUB') do |author|
      @options[:epub_author] = author
    end

    opts.on('--[no-]save-to-icloud', 'Save EPUB to iCloud Drive (default: true)') do |save|
      @options[:save_to_icloud] = save
    end

    opts.on('--[no-]use-llm', 'Use local LLM for intelligent button detection (default: false)') do |llm|
      @options[:use_llm] = llm
    end
  end

  def default_options
    super.merge({
                  timeout: 30,
                  headless: true,
                  wait: 3,
                  max_clicks: 5,
                  use_javascript: false,
                  cache_enabled: true,
                  articles_only: false,
                  create_epub: false,
                  save_to_icloud: true,
                  use_llm: false
                })
  end

  def validate!
    if @args.empty?
      log_error('Please provide a URL to scrape')
      log_info('Usage: website-epub <url>')
      exit 1
    end

    url = @args.first
    unless url.match?(%r{\Ahttps?://})
      log_error('URL must start with http:// or https://')
      exit 1
    end

    super
  end

  def show_examples
    puts 'Examples:'
    puts "  #{script_name} https://nav.al                    # Extract all URLs"
    puts "  #{script_name} https://nav.al --articles-only    # Only articles"
    puts "  #{script_name} https://blog.com --javascript     # Use JavaScript for fetching"
    puts "  #{script_name} https://site.com --no-cache       # Disable caching"
    puts "  #{script_name} https://nav.al --max-clicks 10    # More aggressive clicking"
    puts "  #{script_name} https://nav.al --use-llm          # Use AI for smart button detection"
    puts "  #{script_name} https://nav.al --create-epub      # Create EPUB from articles"
    puts "  #{script_name} https://nav.al --create-epub --epub-title 'Naval Essays' # Custom title"
    puts "  #{script_name} https://nav.al --create-epub --no-save-to-icloud # Don't save to iCloud"
    puts "  #{script_name} https://blog.com --use-llm --create-epub # AI detection + EPUB creation"
  end

  def run
    log_banner("#{script_emoji} Website EPUB Creator")

    url = @args.first
    log_info("Target URL: #{url}")
    log_info('Configuration:')
    log_info("  • Max clicks: #{@options[:max_clicks]}")
    log_info("  • JavaScript mode: #{@options[:use_javascript]}")
    log_info("  • Cache enabled: #{@options[:cache_enabled]}")
    log_info("  • Articles only: #{@options[:articles_only]}")
    log_info("  • LLM detection: #{@options[:use_llm]}")

    browser_service = nil
    page_fetcher = nil

    begin
      # Step 1: Collect URLs using browser automation
      browser_service = BrowserService.new(url, @options.merge(logger: self, debug: debug?))
      all_urls = browser_service.collect_urls_with_read_more

      if all_urls.empty?
        log_warning('No URLs found on the webpage')
        return
      end

      # Step 2: Filter to same domain only
      domain = extract_domain(url)
      same_domain_urls = all_urls.select { |url_info| same_domain?(url_info[:url], domain) }

      log_info("Domain filtering: #{all_urls.size} total → #{same_domain_urls.size} same-domain URLs")

      if same_domain_urls.empty?
        log_warning('No same-domain URLs found')
        return
      end

      # Step 3: Article detection (if enabled)
      final_urls = same_domain_urls

      if @options[:articles_only]
        puts "\n🔍 Detecting articles..."

        # Initialize page fetcher for article detection
        page_fetcher = PageFetcher.new(
          cache_enabled: @options[:cache_enabled],
          verbose: verbose?
        )

        # Extract just the URLs for analysis
        url_strings = same_domain_urls.map { |url_info| url_info[:url] }

        # Filter to articles only
        article_urls = ArticleDetector.filter_article_urls(
          url_strings,
          page_fetcher,
          javascript: @options[:use_javascript]
        )

        # Convert back to our URL info format
        final_urls = same_domain_urls.select { |url_info| article_urls.include?(url_info[:url]) }

        log_info("Article detection: #{same_domain_urls.size} URLs → #{final_urls.size} articles")

        if final_urls.empty?
          log_warning('No articles detected in the URLs')
          return
        end

        # Store article URLs for EPUB generation
        @article_urls = article_urls
      end

      # Step 4: Create EPUB if requested
      if @options[:create_epub]
        create_epub_from_urls(final_urls)
      else
        # Step 4: Display results
        display_urls(final_urls)
      end

      if verbose?
        puts "\n"
        show_cache_stats(page_fetcher) if page_fetcher
      end
    rescue StandardError => e
      log_error("Failed to process URLs: #{e.message}")
      log_debug("Backtrace: #{e.backtrace.join("\n")}") if debug?
    ensure
      browser_service&.close
      page_fetcher&.close
      log_info('Cleanup completed')
    end

    show_completion('Website processing')
  end

  private

  def create_epub_from_urls(urls)
    return if urls.empty?

    log_section('📚 Creating EPUB')

    # Initialize EPUB generator
    epub_generator = EPUBGenerator.new({
                                         cache_enabled: @options[:cache_enabled],
                                         javascript: @options[:use_javascript],
                                         save_to_icloud: @options[:save_to_icloud],
                                         verbose: verbose?,
                                         debug: debug?,
                                         logger: self
                                       })

    # Generate EPUB from article URLs
    if @options[:articles_only]
      # Use the already filtered article URLs
      epub_path = epub_generator.generate_epub_from_article_urls(
        @article_urls,
        @options[:epub_title],
        @options[:epub_author]
      )
    else
      # Extract URLs for EPUB generation
      url_strings = urls.map { |url_info| url_info[:url] }
      epub_path = epub_generator.generate_epub_from_urls(
        url_strings,
        @options[:epub_title],
        @options[:epub_author]
      )
    end

    if epub_path
      log_success('EPUB created successfully!')
      log_info("EPUB location: #{epub_path}")

      # Show iCloud location if saved to iCloud
      if @options[:save_to_icloud] && epub_generator.icloud_available?
        domain = extract_domain(@args.first)
        nice_domain = domain.gsub('.', '-') if domain
        log_info("Check your iCloud Drive for the EPUB in: WebsiteEPUB/#{nice_domain}/")
      end
    else
      log_error('Failed to create EPUB')
    end
  end

  def extract_domain(url)
    uri = URI(url)
    uri.host
  rescue URI::InvalidURIError
    nil
  end

  def same_domain?(url, domain)
    return false unless url.start_with?('http')

    begin
      uri = URI(url)
      uri.host == domain || uri.host&.end_with?(".#{domain}")
    rescue URI::InvalidURIError
      false
    end
  end

  def show_cache_stats(page_fetcher)
    return unless page_fetcher

    js_stats = page_fetcher.cache_stats(namespace: 'js')
    no_js_stats = page_fetcher.cache_stats(namespace: 'no_js')

    log_section('💾 Cache Statistics')
    puts "JavaScript pages: #{js_stats[:files]} files (#{js_stats[:total_size_mb]} MB)"
    puts "No-JS pages: #{no_js_stats[:files]} files (#{no_js_stats[:total_size_mb]} MB)"
    puts "Cache directory: #{js_stats[:cache_dir]}"
  end

  def display_urls(urls)
    log_section('📋 Final URLs')

    urls.each_with_index do |entry, index|
      puts "#{index + 1}. #{entry[:url]}"

      next unless verbose?

      puts "   📝 Text: #{entry[:text]}" unless entry[:text] == '[No text]'
      puts "   🏷️  Title: #{entry[:title]}" unless entry[:title].to_s.empty?
      puts
    end

    puts
    log_success("Total: #{urls.size} URLs ready for EPUB")
  end
end

# Execute the script
WebsiteEpub.execute if __FILE__ == $0

File: bin/bun-cli/src/scripts/epub/website-epub.ts
import { Script, BaseScript, type Context } from '../../core/decorators/script';
import { EpubGeneratorService } from '../../core/services/EpubGeneratorService';
import { PageFetcherService } from '../../core/services/PageFetcherService';
import { ArticleDetectorService } from '../../core/services/ArticleDetectorService';
import { BrowserService } from '../../core/services/BrowserService';
import { mkdirSync } from 'fs';
import { join } from 'path';
import { format } from 'date-fns';

/**
 * Website EPUB Creator
 *
 * Scrapes a website using browser automation and creates an EPUB from articles
 */
@Script({
  emoji: '🌐',
  tags: ['epub', 'website', 'scraper'],
  args: {
    url: {
      type: 'string',
      position: 0,
      description: 'URL to scrape',
      required: true,
    },
    maxClicks: {
      type: 'number',
      flag: '--max-clicks <count>',
      description: 'Maximum read more clicks',
      default: 5,
    },
    headless: {
      type: 'boolean',
      flag: '--headless',
      description: 'Run browser in headless mode',
      default: true,
    },
    timeout: {
      type: 'number',
      flag: '--timeout <seconds>',
      description: 'Browser timeout in seconds',
      default: 30,
    },
    wait: {
      type: 'number',
      flag: '--wait <seconds>',
      description: 'Wait time for page load in seconds',
      default: 3,
    },
    cacheEnabled: {
      type: 'boolean',
      flag: '--cache',
      description: 'Enable page caching',
      default: true,
    },
    articlesOnly: {
      type: 'boolean',
      flag: '--articles-only',
      description: 'Only process URLs that are detected as articles',
      default: false,
    },
    createEpub: {
      type: 'boolean',
      flag: '--create-epub',
      description: 'Create EPUB file from detected articles',
      default: false,
    },
    epubTitle: {
      type: 'string',
      flag: '--epub-title <title>',
      description: 'Custom title for the EPUB',
    },
    epubAuthor: {
      type: 'string',
      flag: '--epub-author <author>',
      description: 'Custom author for the EPUB',
    },
    saveToIcloud: {
      type: 'boolean',
      flag: '--save-to-icloud',
      description: 'Save EPUB to iCloud Drive',
      default: true,
    },
  },
})
export class WebsiteEpubScript extends BaseScript {
  private browserService!: BrowserService;
  private pageFetcher!: PageFetcherService;
  private articleDetector!: ArticleDetectorService;

  async validate(ctx: Context): Promise<void> {
    const args = ctx.args as any;

    // Validate URL
    const url = args.url;
    if (!url || !url.match(/^https?:\/\//)) {
      this.logger.error('URL must start with http:// or https://');
      process.exit(1);
    }

    // Initialize services
    this.browserService = new BrowserService({ logger: this.logger });
    this.pageFetcher = new PageFetcherService({
      logger: this.logger,
      rateLimitDelay: 1000,
      maxConcurrent: 5,
    });
    this.articleDetector = new ArticleDetectorService({
      logger: this.logger,
    });
  }

  async execute(ctx: Context): Promise<void> {
    const args = ctx.args as any;

    this.logger.info('🌐 Website EPUB Creator');
    this.logger.info(`Target URL: ${args.url}`);
    this.logger.info('Configuration:');
    this.logger.info(`  • Max clicks: ${args.maxClicks}`);
    this.logger.info(`  • Cache enabled: ${args.cacheEnabled}`);
    this.logger.info(`  • Articles only: ${args.articlesOnly}`);

    try {
      // Step 1: Initialize browser service
      const browserAvailable = await this.browserService.initialize();

      if (!browserAvailable) {
        this.logger.warn('Browser service not available, falling back to simple HTTP fetching');
      }

      // Step 2: Collect URLs
      const allUrls = await this.collectUrls(args.url, browserAvailable, args);

      if (allUrls.length === 0) {
        this.logger.warn('No URLs found on the webpage');
        return;
      }

      // Step 3: Filter to same domain
      const domain = this.extractDomain(args.url);
      const sameDomainUrls = allUrls.filter((url) => this.isSameDomain(url, domain));

      this.logger.info(
        `Domain filtering: ${allUrls.length} total → ${sameDomainUrls.length} same-domain URLs`
      );

      if (sameDomainUrls.length === 0) {
        this.logger.warn('No same-domain URLs found');
        return;
      }

      // Step 4: Article detection (if enabled)
      let finalUrls = sameDomainUrls;

      if (args.articlesOnly) {
        this.logger.info('🔍 Detecting articles...');
        finalUrls = await this.filterArticles(sameDomainUrls);

        this.logger.info(
          `Article detection: ${sameDomainUrls.length} URLs → ${finalUrls.length} articles`
        );

        if (finalUrls.length === 0) {
          this.logger.warn('No articles detected in the URLs');
          return;
        }
      }

      // Step 5: Create EPUB or display URLs
      if (args.createEpub) {
        await this.createEpubFromUrls(finalUrls, ctx);
      } else {
        this.displayUrls(finalUrls);
      }
    } catch (error: any) {
      this.logger.error(`Failed to process URLs: ${error.message}`);
      throw error;
    } finally {
      // Cleanup
      await this.browserService.close();
      this.logger.debug('Cleanup completed');
    }

    this.logger.success('Website processing completed');
  }

  /**
   * Collect URLs from the website
   */
  private async collectUrls(
    url: string,
    useBrowser: boolean,
    args: any
  ): Promise<string[]> {
    if (useBrowser) {
      return await this.collectUrlsWithBrowser(url, args);
    } else {
      return await this.collectUrlsWithHttp(url);
    }
  }

  /**
   * Collect URLs using browser automation
   */
  private async collectUrlsWithBrowser(url: string, args: any): Promise<string[]> {
    this.logger.info('📖 Collecting URLs with browser automation...');

    const urls = new Set<string>();

    // Navigate to the page
    const page = await this.browserService.newPage();

    try {
      await page.goto(url, {
        waitUntil: 'networkidle0',
        timeout: args.timeout * 1000,
      });

      // Wait for initial page load
      await page.waitForTimeout(args.wait * 1000);

      // Click "Read More" buttons
      for (let i = 0; i < args.maxClicks; i++) {
        this.logger.debug(`Click iteration ${i + 1}/${args.maxClicks}`);

        // Try to find and click "read more" type buttons
        const clicked = await this.tryClickReadMore(page);

        if (!clicked) {
          this.logger.debug('No more read more buttons found');
          break;
        }

        // Wait for content to load
        await page.waitForTimeout(args.wait * 1000);
      }

      // Extract all links
      const links = await this.browserService.extractLinks(url);
      links.forEach((link) => urls.add(link.href));

      this.logger.success(`Collected ${urls.size} URLs from browser`);
    } finally {
      await page.close();
    }

    return Array.from(urls);
  }

  /**
   * Try to click "read more" buttons
   */
  private async tryClickReadMore(page: any): Promise<boolean> {
    try {
      // Common selectors for "read more" buttons
      const selectors = [
        'button:has-text("Read More")',
        'button:has-text("Load More")',
        'button:has-text("Show More")',
        'a:has-text("Read More")',
        'a:has-text("Load More")',
        '.load-more',
        '.read-more',
        '[data-load-more]',
      ];

      for (const selector of selectors) {
        try {
          const element = await page.$(selector);
          if (element) {
            await element.click();
            return true;
          }
        } catch {
          // Continue to next selector
        }
      }

      return false;
    } catch {
      return false;
    }
  }

  /**
   * Collect URLs using simple HTTP fetching
   */
  private async collectUrlsWithHttp(url: string): Promise<string[]> {
    this.logger.info('📖 Collecting URLs with HTTP fetching...');

    const page = await this.pageFetcher.fetchPage(url);
    const links = this.articleDetector.extractLinks(page.content);

    const urls = links.map((link) => this.resolveUrl(link.href, url));

    this.logger.success(`Collected ${urls.length} URLs from HTTP`);
    return urls;
  }

  /**
   * Filter URLs to articles only
   */
  private async filterArticles(urls: string[]): Promise<string[]> {
    const articles: string[] = [];

    for (const url of urls) {
      try {
        const page = await this.pageFetcher.fetchPage(url);
        const isArticle = this.articleDetector.isProbablyArticle(page.content);

        if (isArticle) {
          articles.push(url);
        }
      } catch (error: any) {
        this.logger.debug(`Failed to check article status for ${url}: ${error.message}`);
      }
    }

    return articles;
  }

  /**
   * Create EPUB from URLs
   */
  private async createEpubFromUrls(urls: string[], ctx: Context): Promise<void> {
    const args = ctx.args as any;

    this.logger.info('📚 Creating EPUB');

    // Fetch all pages
    this.logger.info(`Fetching ${urls.length} pages...`);
    const pages = await this.pageFetcher.fetchPages(urls);

    if (pages.length === 0) {
      this.logger.error('Failed to fetch any pages');
      return;
    }

    this.logger.success(`Fetched ${pages.length}/${urls.length} pages`);

    // Extract articles
    this.logger.info('Extracting article content...');
    const articles = this.articleDetector.extractArticles(
      pages.map((p) => ({ html: p.content, url: p.url }))
    );

    if (articles.length === 0) {
      this.logger.error('Failed to extract any articles');
      return;
    }

    this.logger.success(`Extracted ${articles.length} articles`);

    // Generate EPUB
    const domain = this.extractDomain(args.url);
    const title = args.epubTitle || `${domain} Articles`;
    const author = args.epubAuthor || domain;

    const epubGenerator = new EpubGeneratorService(
      {
        title,
        author,
        description: `Articles from ${domain}, created on ${format(new Date(), 'MMMM d, yyyy')}`,
      },
      this.logger
    );

    // Add chapters
    articles.forEach((article) => {
      epubGenerator.addChapter({
        title: article.title,
        content: article.content,
        author: article.author,
        url: article.url,
        date: article.publishedTime ? new Date(article.publishedTime) : undefined,
      });
    });

    // Generate EPUB file
    const outputPath = this.getOutputPath(domain, args.saveToIcloud);
    await epubGenerator.generate(outputPath);

    this.logger.success('EPUB created successfully!');
    this.logger.info(`EPUB location: ${outputPath}`);

    if (args.saveToIcloud) {
      const niceDomain = domain.replace(/\./g, '-');
      this.logger.info(`Check your iCloud Drive for the EPUB in: WebsiteEPUB/${niceDomain}/`);
    }
  }

  /**
   * Display collected URLs
   */
  private displayUrls(urls: string[]): void {
    this.logger.info('📋 Final URLs\n');

    urls.forEach((url, index) => {
      console.log(`${index + 1}. ${url}`);
    });

    console.log();
    this.logger.success(`Total: ${urls.length} URLs`);
  }

  /**
   * Extract domain from URL
   */
  private extractDomain(url: string): string {
    try {
      const urlObj = new URL(url);
      return urlObj.hostname;
    } catch {
      return 'unknown';
    }
  }

  /**
   * Check if two URLs are from the same domain
   */
  private isSameDomain(url: string, domain: string): boolean {
    if (!url.startsWith('http')) return false;

    try {
      const urlObj = new URL(url);
      return urlObj.hostname === domain || urlObj.hostname.endsWith(`.${domain}`);
    } catch {
      return false;
    }
  }

  /**
   * Resolve relative URL to absolute URL
   */
  private resolveUrl(href: string, baseUrl: string): string {
    try {
      return new URL(href, baseUrl).href;
    } catch {
      return href;
    }
  }

  /**
   * Get output path for EPUB
   */
  private getOutputPath(domain: string, saveToIcloud: boolean): string {
    const niceDomain = domain.replace(/\./g, '-');
    const filename = `${niceDomain}_${format(new Date(), 'yyyy-MM-dd')}.epub`;

    if (saveToIcloud) {
      const icloudPath = join(
        process.env.HOME || '',
        `Library/Mobile Documents/com~apple~CloudDocs/WebsiteEPUB/${niceDomain}`
      );
      mkdirSync(icloudPath, { recursive: true });
      return join(icloudPath, filename);
    } else {
      const downloadsPath = join(process.env.HOME || '', 'Downloads');
      return join(downloadsPath, filename);
    }
  }
}
